### PR TITLE
Mark `node:` as external for Node.js compatibility

### PR DESCRIPTION
**Pull request by chientrm**
### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

## Explain

Last time I asked to add polyfill for `Buffer` API. That's no longer neccessary because Cloudflare added a bunch of nodejs APIs to Workers. They are all prefixed by `node:`.

Example: 

```ts
import { Buffer } from 'node:buffer';
```

These APIs are enabled via compatibility_flags `nodejs_compat` and work for Workers, Pages and `wrangler`.

The APIs is here: [Node.js compatibility](https://developers.cloudflare.com/workers/runtime-apis/nodejs/)

### CODE REVIEW

**Comment by ghostdevv** [comment]
I'm for this change but I think it may make sense having it behind some sort of config option, maybe `nodeCompat`? That way we can document what this means and not suddenly start allowing node packages for people who don't know what that means on Cloudflare.

**Comment by chientrm** [comment]
> I'm for this change but I think it may make sense having it behind some sort of config option, maybe `nodeCompat`? That way we can document what this means and not suddenly start allowing node packages for people who don't know what that means on Cloudflare.

I committed the suggested solution yesterday, not sure GitHub will notify the mods so that I quote your reply here.

**Comment by ghostdevv** [comment]
We should also add notes to the docs about this :pray: We'll also will need some opinions from other maintainers before we can merge

**Comment by benmccann** [comment]
We generally try to avoid adding options, so unless there's a scenario where we don't want to do it, I would suggest we should always do it.

However, I'm not sure whether it's enough? @cimnine reported in https://github.com/sveltejs/kit/pull/10521 that they had to use `alias` to actually address an issue using a Node.js module. Is there some kind of guidance we could provide users as to when `external` is sufficient and when `alias` would be required?

**Comment by ghostdevv** [comment]
> However, I'm not sure whether it's enough? @cimnine reported in https://github.com/sveltejs/kit/pull/10521 that they had to use alias to actually address an issue using a Node.js module. Is there some kind of guidance we could provide users as to when external is sufficient and when alias would be required?

I'm still not 100% sure what the decision is re: various esbuild options being passable to the adapters - regardless this might be a nice change to reduce friction even if we do end up merging one of the esbuild option related prs - WDYT? @benmccann

**Review by dummdidumm — COMMENTED** [review]
#10521 is related - if we decide to allow `external` configuration (which it seems there's now enough evidence that we should) then we don't need this.

**Comment by JustMrMendez** [comment]
any updates on this? can't get postgres working on cloudflare and this might be related

**Comment by chientrm** [comment]
> any updates on this? can't get postgres working on cloudflare and this might be related

You can try my adapter while waiting [@chientrm/adapter-cloudflare](https://www.npmjs.com/package/@chientrm/adapter-cloudflare).

Remember to add flag `nodejs_compat` to your Workers or Pages.

**Comment by superboss224** [comment]
Hello, will the fix be merged anytime soon ?

**Comment by carstenjaksch** [comment]
> > any updates on this? can't get postgres working on cloudflare and this might be related
> 
> You can try my adapter while waiting [@chientrm/adapter-cloudflare](https://www.npmjs.com/package/@chientrm/adapter-cloudflare).
> 
> Remember to add flag `nodejs_compat` to your Workers or Pages, and set `adapter: adapter({ nodeCompat: true })`.

Unfortunately, that doesn't work for me. I get these errors on Cloudflare Pages during build:

```
The package "stream" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.
```

The same with a lot other Node packages in `nodemailer`.

**Comment by chientrm** [comment]
> > > any updates on this? can't get postgres working on cloudflare and this might be related
> > 
> > 
> > You can try my adapter while waiting [@chientrm/adapter-cloudflare](https://www.npmjs.com/package/@chientrm/adapter-cloudflare).
> > Remember to add flag `nodejs_compat` to your Workers or Pages, and set `adapter: adapter({ nodeCompat: true })`.
> 
> Unfortunately, that doesn't work for me. I get these errors on Cloudflare Pages during build:
> 
> ```
> The package "stream" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.
> ```
> 
> The same with a lot other Node packages in `nodemailer`.

how do you import `stream`?

**Comment by carstenjaksch** [comment]
> > > > any updates on this? can't get postgres working on cloudflare and this might be related
> > > 
> > > 
> > > You can try my adapter while waiting [@chientrm/adapter-cloudflare](https://www.npmjs.com/package/@chientrm/adapter-cloudflare).
> > > Remember to add flag `nodejs_compat` to your Workers or Pages, and set `adapter: adapter({ nodeCompat: true })`.
> > 
> > 
> > Unfortunately, that doesn't work for me. I get these errors on Cloudflare Pages during build:
> > ```
> > The package "stream" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The same with a lot other Node packages in `nodemailer`.
> 
> how do you import `stream`?

I do not import it, but `nodemailer` does.

**Comment by chientrm** [comment]
you can fork `nodemailer` and make it import `node:stream` instead of `stream`. Docs here: https://github.com/sveltejs/kit/pull/10544#issuecomment-1870098868

**Comment by chientrm** [comment]
> > > > > any updates on this? can't get postgres working on cloudflare and this might be related
> > > > 
> > > > 
> > > > You can try my adapter while waiting [@chientrm/adapter-cloudflare](https://www.npmjs.com/package/@chientrm/adapter-cloudflare).
> > > > Remember to add flag `nodejs_compat` to your Workers or Pages, and set `adapter: adapter({ nodeCompat: true })`.
> > > 
> > > 
> > > Unfortunately, that doesn't work for me. I get these errors on Cloudflare Pages during build:
> > > ```
> > > The package "stream" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The same with a lot other Node packages in `nodemailer`.
> > 
> > 
> > how do you import `stream`?
> 
> I do not import it, but `nodemailer` does.

I updated my `@chientrm/adapter-cloudflare` recently, it enable node compat by default, no need `adapter: adapter({ nodeCompat: true })`. At this point I don't think the team's gonna merge this pull request. You may aswell fork your own adapter and publish for your personal use.

File: documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md
Comment (benmccann) [inline documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md:94]:
```suggestion
You will also need to [enable Node.js with Workers](https://developers.cloudflare.com/workers/runtime-apis/nodejs/#enable-nodejs-with-workers).
```

File: documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md
Comment (benmccann) [inline documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md:89]: why pass it in if it's already specified in the `wrangler.toml`? https://developers.cloudflare.com/workers/runtime-apis/nodejs/#enable-nodejs-with-workers

I think it'd be nicer to read it from the `wrangler.toml` to avoid making the user specify it twice

**Review by chientrm — COMMENTED** [review]
Hmm, I don't know how to merge this commit lol.

File: documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md
Comment (chientrm) [reply documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md:89]: When deploy to Cloudflare Pages, the file `wrangler.toml' is not available.

Comment (benmccann) [reply documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md:89]: Yes, but this is the Cloudflare Workers adapter :grin:

File: packages/adapter-cloudflare-workers/index.js
Comment (benmccann) [inline packages/adapter-cloudflare-workers/index.js:75]: Whitelisting `node:*` may be too aggressive. I think if we do that and someone uses an API like `node:fs` then they won't find out their app is broken until they go to deploy it and it'd be better for that to occur during build. How about we switch it to just the list of APIs that Cloudflare supports with their node compatibility?

'node:assert', 'node:async_hooks', 'node:buffer', 'node:crypto', node:diagnostics_channel', node:events', node:path', node:process', node:stream', node:string_decoder', node:util'

**Comment by benmccann** [comment]
> Hmm, I don't know how to merge this commit lol.

I merged it via the GitHub UI for you

File: packages/adapter-cloudflare-workers/index.js
Comment (chientrm) [reply packages/adapter-cloudflare-workers/index.js:75]: Yep that sounds legit. Cloudflare node_compat doesn't have modules like `fs`, ...

File: documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md
Comment (chientrm) [reply documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md:89]: Oke, I'm going to make it read the `wrangler.toml`, not sure how could I do that. Maybe just `string.includes('node_compat')`? 🤔

File: packages/adapter-cloudflare-workers/index.js
Comment (chientrm) [reply packages/adapter-cloudflare-workers/index.js]: Is fixed in my latest commit

File: documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md
Comment (chientrm) [reply documentation/docs/25-build-and-deploy/70-adapter-cloudflare-workers.md:89]: oh, there is a `toml.parse` already.

File: packages/adapter-cloudflare-workers/index.js
Comment (chientrm) [reply packages/adapter-cloudflare-workers/index.js]: May as well I remove `AdapterOptions.nodeCompat` and always included those externals? 🤔

File: documentation/docs/25-build-and-deploy/60-adapter-cloudflare.md
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/60-adapter-cloudflare.md:32]:
```suggestion
			}
```

File: documentation/docs/25-build-and-deploy/60-adapter-cloudflare.md
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/60-adapter-cloudflare.md:50]:
```suggestion
```

File: packages/adapter-cloudflare-workers/index.js
Comment (Rich-Harris) [inline packages/adapter-cloudflare-workers/index.js:19]: Why did this change? The old code provided a default value for `config`; the new code provides a default value for` options` but _not_ for `config` specifically. As such, `options.config` could be undefined

File: packages/adapter-cloudflare/index.d.ts
Comment (Rich-Harris) [inline packages/adapter-cloudflare/index.d.ts:50]:
```suggestion
```

File: packages/adapter-cloudflare/index.js
Comment (Rich-Harris) [inline packages/adapter-cloudflare/index.js:71]:
```suggestion
			const external = [
				'cloudflare:*',
				'node:assert',
				'node:async_hooks',
				'node:buffer',
				'node:crypto',
				'node:diagnostics_channel',
				'node:events',
				'node:path',
				'node:process',
				'node:stream',
				'node:string_decoder',
				'node:util'
			];
```

**Review by Rich-Harris — COMMENTED** [review]
Thank you!

We should remove the `nodeCompat` option — there's no real value in it since it doesn't reflect or determine the project's compatibility flag. Better to only have to set it in one place.

Though it's probably worth generating a useful error when using unsupported modules like `node:fs`, or supported-but-unprefixed modules like `path` — will do that in a follow-up PR

File: packages/adapter-cloudflare-workers/index.js
Comment (Rich-Harris) [reply packages/adapter-cloudflare-workers/index.js:19]: reverting the change

```suggestion
export default function ({ config = 'wrangler.toml' } = {}) {
```

File: packages/adapter-cloudflare-workers/index.js
Comment (Rich-Harris) [inline packages/adapter-cloudflare-workers/index.js:24]:
```suggestion
			const { main, site, compatibility_flags } = validate_config(builder, config);
```

**Review by Rich-Harris — APPROVED** [review]

File: .changeset/pretty-geese-drum.md
Comment (Rich-Harris) [inline .changeset/pretty-geese-drum.md:6]:
```suggestion
---
'@sveltejs/adapter-cloudflare-workers': minor
'@sveltejs/adapter-cloudflare': minor
---

feat: Add Node.js compatibility
```

**Comment by mgibbs189** [comment]
@chientrm @benmccann @Rich-Harris Thanks! 🎉

**Comment by Hebilicious** [comment]
@Rich-Harris While things are better, still are still un-usable in some cases :/ 

When bundling a project with wrangler and the node_compat flag, several things happens 

- https://github.com/cloudflare/workers-sdk/blob/main/packages/wrangler/src/deployment-bundle/esbuild-plugins/nodejs-compat.ts this plugin flags all node modules are externals.

- These polyfills https://github.com/cloudflare/workers-sdk/blob/ef0642796dbe17b30fd7b83ccfd3efc651ce0a1a/packages/wrangler/src/deployment-bundle/bundle.ts#L334 are injected

This allows cloudflare workers and cloudflare pages projects to correctly build if there's a supported library such as "pg" that doesn't use "node:x" specifiers and use incompatible APIs.

Note that there's also an issue with `wrangler deploy` where this is not possible to specify node_compat: true, therefore the project must be built manually, then deployed.

To recap, I'm using a forked version of the cloudflare adapter to treat everything as externals and to inject the polyfills.
Another gotcha is that the nodejs compatibility flag is *not compatible* with the polyfills while building with wrangler ... But it's still possible to build manually and enable the flag.

I'm not sure what would be a possible solution, but it would be nice to enable esbuild custom settings for the cloudflare adapters.

**Comment by g-wozniak** [comment]
I am getting exceptions for every single package I am trying to use. 
```
{
  "outcome": "exception",
  "scriptName": "pages-worker--1989843-preview",
  "diagnosticsChannelEvents": [],
  "exceptions": [
    {
      "name": "Error",
      "message": "Dynamic require of \"node:crypto\" is not supported",
      "timestamp": 1708693796319
    }
  ],
}
```

Using:
- "@sveltejs/kit": "^2.5.0"
- "@sveltejs/adapter-cloudflare": "^4.1.0"

I think, inside 'hooks.server.ts', got a custom handle which performs:
```
import {V4} from 'paseto'
...
await V4.verify(sessionToken, secretKey)
```

and I suspect that's the issue.

Got "nodejs_compat" flag enabled but it does not help.

**Comment by g-wozniak** [comment]
Screenshot 2024-02-23 at 13 45 08

Yep, I even changed it to something else to see if the build fails and it does.

**Comment by g-wozniak** [comment]
According to the [documentation](https://developers.cloudflare.com/workers/wrangler/configuration/#use-runtime-apis-directly) the flag name is: "nodejs_compat"

**Comment by absktoday** [comment]
Is this supposed to work for Pages Adapter or do I have to use Worker adapter, because I am still getting the Could not resolve "crypto" or Could not resolve "node:crypto".

**Comment by devinellis** [comment]
Same question as @absktoday but for a dependency that is attempting to import `url`. [Cloudflare states this is supported in Pages](https://developers.cloudflare.com/workers/runtime-apis/nodejs/#get-started), and yet the adapter throws an error. I am using Cloudflare Pages which does not use wrangler.toml (as far as I know?). I created one anyways to see if it would "trick" the adapter but it did not work. My workaround is to fork the dependency and remove those imports :(

**Comment by ghostdevv** [comment]
Is the error at runtime on Cloudflare pages, or at buildtime either locally or on the CF pages builder @devinellis? What imports is the package using `url` or `node:url`? Could you share a minimal reproduction?

**Comment by devinellis** [comment]
Sorry for the lack of detail:
- The package is doing `require('url')`. Commenting these lines out in a fork made the problem go away.
- It fails both locally and in the CF Pages builder.
- Setting `nodejs_compat` in wrangler.toml or in the Pages Dashboard didn't seem to make a difference.

**Note** if I am reading the [CF docs](https://developers.cloudflare.com/workers/configuration/compatibility-dates/#nodejs-compatibility-flag) right, `nodejs_compat` no longer needs the `node:` prefix on package names, as long as your `compatibility_date` > 2024-09-23

**Pull request merged**
